contracts/walrus/sources/blob_manager.move

```diff
@@ -0,0 +1,833 @@
+    let manager_uid = object::new(ctx);
+
+    let manager = BlobManager {
+        id: manager_uid,
+        storage: blob_storage::new_unified_blob_storage(initial_storage, ctx),
+        coin_stash: coin_stash::new(),
+        // Default policy: extend within 1 epoch of expiry, max 10 epochs per extension.
+        extension_policy: extension_policy::constrained(1, 10),
+        // Default tip policy: 1000 MIST (0.001 SUI) fixed amount.
+        tip_policy: tip_policy::new_fixed_amount(1000),
+        grace_period_epochs,
+    };
+
+    // Get the ObjectID from the constructed manager object.
+    let manager_object_id = object::id(&manager);
+
+    // Create an Admin capability with fund_manager permissions.
+    let cap = BlobManagerCap {
+        id: object::new(ctx),
+        manager_id: manager_object_id,
+        is_admin: true,
+        fund_manager: true,
+    };
+
+    // Emit creation event.
+    events::emit_blob_manager_created(
+        current_epoch,
+        manager_object_id,
+        end_epoch,
+        grace_period_epochs,
+    );
+
+    // BlobManager is designed to be a shared object.
+    transfer::share_object(manager);
+
+    cap
+}
+
+// === Capability Operations ===
+
+/// Creates a new capability for the BlobManager.
+/// Only Admin capability can create new capabilities.
+/// If the creating cap has fund_manager = true, they can create new caps with fund_manager = true.
+/// Returns the newly created capability (caller/PTB handles transfer).
+public fun create_cap(
+    self: &BlobManager,
+    cap: &BlobManagerCap,
+    is_admin: bool,
+    fund_manager: bool,
+    ctx: &mut TxContext,
+): BlobManagerCap {
+    // Verify the capability matches this BlobManager.
+    check_cap(self, cap);
+
+    // Ensure the caller has Admin capability.
+    ensure_admin(cap);
+
+    // Only caps with fund_manager = true can create new fund_manager caps.
+    assert!(!fund_manager || cap.fund_manager, ERequiresFundManager);
+
+    BlobManagerCap {
+        id: object::new(ctx),
+        manager_id: object::id(self),
+        is_admin,
+        fund_manager,
+    }
+}
+
+/// Returns the manager ID from a capability.
+public fun cap_manager_id(cap: &BlobManagerCap): ID {
+    cap.manager_id
+}
+
+/// Checks if the capability is an Admin capability.
+public fun is_admin_cap(cap: &BlobManagerCap): bool {
+    cap.is_admin
+}
+
+/// Checks if the capability has fund manager permissions.
+public fun is_fund_manager_cap(cap: &BlobManagerCap): bool {
+    cap.fund_manager
+}
+
+/// Checks that the given BlobManagerCap matches the BlobManager.
+fun check_cap(self: &BlobManager, cap: &BlobManagerCap) {
+    let manager_id = cap_manager_id(cap);
+    assert!(object::id(self) == manager_id, EInvalidBlobManagerCap);
+}
+
+/// Ensures the capability is an Admin capability.
+fun ensure_admin(cap: &BlobManagerCap) {
+    assert!(is_admin_cap(cap), ERequiresAdminCap);
+}
+
+/// Calculates grace period based on duration.
```

Comment:
As before, not sure that grace periods are a must, so reducing code in this large PR may be preferable.

Reply:
grace-period removed